cli/templatecheckout.go

```diff
@@ -0,0 +1,85 @@
+package cli
+
+import (
+	"archive/tar"
+	"bytes"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+
+	"github.com/spf13/cobra"
+	"golang.org/x/xerrors"
+
+	"github.com/coder/coder/cli/cliui"
+)
+
+func TarBytesToTree(destination string, raw []byte) error {
+	err := os.Mkdir(destination, 0700)
+
+	archiveReader := tar.NewReader(bytes.NewReader(raw))
+	hdr, err := archiveReader.Next()
+	for err != io.EOF {
+		if hdr == nil { //	some blog post indicated this could happen sometimes
+			continue
+		}
+		filename := filepath.FromSlash(fmt.Sprintf("%s/%s", destination, hdr.Name))
+		switch hdr.Typeflag {
+		case tar.TypeDir:
+			err = os.Mkdir(filename, 0700)
+			if err != nil {
+				return xerrors.Errorf("unable to check out template directory: %w", err)
+			}
+		case tar.TypeReg:
+			f, err := os.OpenFile(filename, os.O_CREATE|os.O_WRONLY, 0600)
+			if err != nil {
+				return xerrors.Errorf("unable to create template file: %w", err)
+			}
+
+			_, err = io.Copy(f, archiveReader)
+			if err != nil {
+				f.Close() // is this necessary?
+				return xerrors.Errorf("error writing template file: %w", err)
+			}
+			f.Close()
+		}
+
+		hdr, err = archiveReader.Next()
+	}
+	return nil
+}
+
+func templateCheckout() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "checkout <template name> [destination]",
```

Comment:
What is the rationale to make this a separate command from `pull`? I feel as though it does essentially the same thing, but `--extract` may be clearer to a caller.

I'm concerned of the `checkout` name, because of the association with `git`.

I'd propose a UX like:

```bash
$ coder templates pull <name>
Warning: Binary output can mess up your terminal. Use "--extract -" to tell
Warning: coder to output it to your terminal anyway, or consider "--extract
Warning: <FOLDER>" to extract to a directory.
```
This output is taken verbatim from `curl`.

Reply:
Well, one reason to make it  separate command is because I think the more common thing will be to download into an unpacked tree, so I don't want people to always have to type `--extract`.  I'm open to different names, but I chose `checkout` precisely because of the similarity with version control systems.

Reply:
`checkout` gives the appearance that it's _checking out_ a ref if you're familiar with Git. I don't think it'd be synonymous with cloning a repository, which this does.

I think extract is better, but having two commands I think is excessively verbose for what this does. It's not a big deal to type `--extract`, and I'm fine to swap the default behavior to extract to a directory too.

Reply:
Okay.  I'll swap the behaviors and add a `--archive` flag.